File: public/code/smc__eq_sweep.py
"""
Equal Highs/Lows Sweep (EQH/EQL)
=================================
İki veya daha fazla yakın seviyede high/low oluşumu = likidite havuzu.
Bu seviyenin süpürülmesi (wick ile aşılıp dönüş) = entry sinyali.
"""
import sys, os
sys.path.append(os.path.join(os.path.dirname(__file__), "..", "..", ".."))
import pandas as pd
import numpy as np
from common import make_signal, atr, swing_highs_lows


def detect_eq_sweep_signal(
    df: pd.DataFrame,
    tolerance_atr: float = 0.15,
    swing_lb: int = 5,
) -> dict:
    if len(df) < 50:
        return make_signal(reason="Insufficient data")

    sh, sl = swing_highs_lows(df, swing_lb)
    current_atr = atr(df, 14).iloc[-1]
    tolerance = current_atr * tolerance_atr

    # Son 20 swing içinde EQH/EQL ara
    recent_highs = sh.dropna().tail(10).values
    recent_lows = sl.dropna().tail(10).values

    eqh_level = None
    eql_level = None

    # EQH: birbirine yakın 2+ high
    for i in range(len(recent_highs) - 1):
        for j in range(i + 1, len(recent_highs)):
            if abs(recent_highs[i] - recent_highs[j]) <= tolerance:
                eqh_level = max(recent_highs[i], recent_highs[j])
                break

    for i in range(len(recent_lows) - 1):
        for j in range(i + 1, len(recent_lows)):
            if abs(recent_lows[i] - recent_lows[j]) <= tolerance:
                eql_level = min(recent_lows[i], recent_lows[j])
                break

    current = df.iloc[-1]
    prev = df.iloc[-2]

    # Bullish sweep: EQL süpürüldü ve dönüş var
    if eql_level is not None and prev["low"] < eql_level and prev["close"] > eql_level \
       and current["close"] > current["open"]:
        entry = current["close"]
        stop_loss = prev["low"] - 0.3 * current_atr
        risk = entry - stop_loss
        return make_signal(
            signal="long",
            entry=entry,
            stop_loss=stop_loss,
            take_profit=[entry + risk * 2, entry + risk * 3, entry + risk * 5],
            confidence=0.76,
            reason=f"EQL sweep at {eql_level:.4f}",
            metadata={"eql": eql_level},
        )

    # Bearish sweep: EQH süpürüldü
    if eqh_level is not None and prev["high"] > eqh_level and prev["close"] < eqh_level \
       and current["close"] < current["open"]:
        entry = current["close"]
        stop_loss = prev["high"] + 0.3 * current_atr
        risk = stop_loss - entry
        return make_signal(
            signal="short",
            entry=entry,
            stop_loss=stop_loss,
            take_profit=[entry - risk * 2, entry - risk * 3, entry - risk * 5],
            confidence=0.76,
            reason=f"EQH sweep at {eqh_level:.4f}",
            metadata={"eqh": eqh_level},
        )

    return make_signal(reason="No EQH/EQL sweep")

File: public/code/smc__eq_sweep.ts
/**
 * Equal Highs/Lows Sweep Strategy
 */
import { Candle, Signal, makeSignal, atr, swingHighs, swingLows } from "../../common";

export function detectEqSweepSignal(
  candles: Candle[],
  opts: { toleranceAtr?: number; swingLb?: number } = {}
): Signal {
  const { toleranceAtr = 0.15, swingLb = 5 } = opts;
  if (candles.length < 50) return makeSignal({ reason: "Insufficient data" });

  const atrValues = atr(candles, 14);
  const currentAtr = atrValues[atrValues.length - 1];
  const tolerance = currentAtr * toleranceAtr;

  const recentHighs = swingHighs(candles, swingLb)
    .filter((v): v is number => v !== null).slice(-10);
  const recentLows = swingLows(candles, swingLb)
    .filter((v): v is number => v !== null).slice(-10);

  let eqhLevel: number | null = null;
  let eqlLevel: number | null = null;

  for (let i = 0; i < recentHighs.length - 1; i++) {
    for (let j = i + 1; j < recentHighs.length; j++) {
      if (Math.abs(recentHighs[i] - recentHighs[j]) <= tolerance) {
        eqhLevel = Math.max(recentHighs[i], recentHighs[j]);
        break;
      }
    }
  }
  for (let i = 0; i < recentLows.length - 1; i++) {
    for (let j = i + 1; j < recentLows.length; j++) {
      if (Math.abs(recentLows[i] - recentLows[j]) <= tolerance) {
        eqlLevel = Math.min(recentLows[i], recentLows[j]);
        break;
      }
    }
  }

  const current = candles[candles.length - 1];
  const prev = candles[candles.length - 2];

  if (eqlLevel !== null && prev.low < eqlLevel && prev.close > eqlLevel && current.close > current.open) {
    const entry = current.close;
    const stopLoss = prev.low - 0.3 * currentAtr;
    const risk = entry - stopLoss;
    return makeSignal({
      signal: "long", entry, stop_loss: stopLoss,
      take_profit: [entry + risk * 2, entry + risk * 3, entry + risk * 5],
      confidence: 0.76, reason: `EQL sweep at ${eqlLevel.toFixed(4)}`,
      timestamp: current.timestamp, metadata: { eql: eqlLevel },
    });
  }

  if (eqhLevel !== null && prev.high > eqhLevel && prev.close < eqhLevel && current.close < current.open) {
    const entry = current.close;
    const stopLoss = prev.high + 0.3 * currentAtr;
    const risk = stopLoss - entry;
    return makeSignal({
      signal: "short", entry, stop_loss: stopLoss,
      take_profit: [entry - risk * 2, entry - risk * 3, entry - risk * 5],
      confidence: 0.76, reason: `EQH sweep at ${eqhLevel.toFixed(4)}`,
      timestamp: current.timestamp, metadata: { eqh: eqhLevel },
    });
  }

  return makeSignal({ reason: "No EQH/EQL sweep" });
}
